File: pufferlib/ocean/bvr/bvr.c
// Standalone C demo for bvr environment
// Compile using: ./scripts/build_ocean.sh bvr [local|fast]
// Run with: ./bvr

#include "bvr.h"
#include "render.h"
#include <stdio.h>
#include <time.h>

void generate_dummy_actions(BvrEnv *env) {
    // Generate random floats in [-1, 1] range
    env->actions[0] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
    env->actions[1] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
    env->actions[2] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
}

void test_performance(int timeout) {
    srand(time(NULL)); // Seed random number generator

    BvrEnv *env = calloc(1, sizeof(BvrEnv));
    env->num_agents = 1;
    init(env);

    size_t obs_size = OBS_DIM * env->framestack;
    size_t act_size = 3;
    env->observations = (float *)calloc(env->num_agents * obs_size, sizeof(float));
    env->actions = (float *)calloc(env->num_agents * act_size, sizeof(float));
    env->rewards = (float *)calloc(env->num_agents, sizeof(float));
    env->terminals = (unsigned char *)calloc(env->num_agents, sizeof(float));

    if (!env->observations || !env->actions || !env->rewards) {
        fprintf(stderr, "ERROR: Failed to allocate memory for demo buffers.\n");
        free(env->observations);
        free(env->actions);
        free(env->rewards);
        free(env->terminals);
        free(env);
        return;
    }

    init(env);
    c_reset(env);

    int start = time(NULL);
    int num_steps = 0;
    while (time(NULL) - start < timeout) {
        generate_dummy_actions(env);
        c_step(env);
        num_steps++;
    }

    int end = time(NULL);
    float sps = (env->num_agents * num_steps) / (float)(end - start);
    printf("Test Environment SPS: %f\n", sps);

    c_close(env);
    free(env->observations);
    free(env->actions);
    free(env->rewards);
    free(env->terminals);
    free(env);
}

int main() {
    srand(time(NULL)); // Seed random number generator

    BvrEnv *env = calloc(1, sizeof(BvrEnv));
    env->num_agents = 1;
    init(env);

    size_t obs_size = OBS_DIM * env->framestack;
    size_t act_size = 3;
    env->observations = (float *)calloc(env->num_agents * obs_size, sizeof(float));
    env->actions = (float *)calloc(env->num_agents * act_size, sizeof(float));
    env->rewards = (float *)calloc(env->num_agents, sizeof(float));
    env->terminals = (unsigned char *)calloc(env->num_agents, sizeof(float));

    if (!env->observations || !env->actions || !env->rewards) {
        fprintf(stderr, "ERROR: Failed to allocate memory for demo buffers.\n");
        free(env->observations);
        free(env->actions);
        free(env->rewards);
        free(env->terminals);
        free(env);
        return 0;
    }

    init(env);
    c_reset(env);
    c_render(env);

    int action_hold = 20;
    generate_dummy_actions(env);
    while (!WindowShouldClose()) {
        if (env->tick % action_hold == 0) {
            generate_dummy_actions(env);
        }
        c_step(env);
        c_render(env);
    }

    c_close(env);
    free(env->observations);
    free(env->actions);
    free(env->rewards);
    free(env->terminals);
    free(env);

    return 0;
}

File: pufferlib/ocean/bvr/bvr.py
import numpy as np
import gymnasium

import pufferlib
from pufferlib.ocean.bvr import binding

class Bvr(pufferlib.PufferEnv):
    def __init__(
        self,
        num_envs=1,
        framestack=1,
        render_mode=None,
        log_interval=128,
        buf=None,
        seed=0,
    ):
        
        self.single_observation_space = gymnasium.spaces.Box(
            low=-1,
            high=1,
            shape=(18 * framestack,),
            dtype=np.float32,
        )

        self.single_action_space = gymnasium.spaces.Box(
            low=-1, high=1, shape=(3,), dtype=np.float32
        )

        self.num_agents = num_envs
        self.framestack = framestack
        self.render_mode = render_mode
        self.log_interval = log_interval
        self.tick = 0

        super().__init__(buf)
        self.actions = self.actions.astype(np.float32)
        self.c_envs = binding.vec_init(
            self.observations,
            self.actions,
            self.rewards,
            self.terminals,
            self.truncations,
            num_envs,
            seed,
            num_agents=1,
            framestack=framestack,
        )

    def reset(self, seed=None):
        self.tick = 0
        seed = 0 if seed is None else seed
        binding.vec_reset(self.c_envs, seed)
        return self.observations, []

    def step(self, actions):
        self.tick += 1

        self.actions[:] = actions
        binding.vec_step(self.c_envs)

        info = []
        if self.tick % self.log_interval == 0:
            info.append(binding.vec_log(self.c_envs))

        return (self.observations, self.rewards, self.terminals, self.truncations, info)

    def render(self):
        binding.vec_render(self.c_envs, 0)

    def close(self):
        binding.vec_close(self.c_envs)

BVR = Bvr

def test_performance(timeout=10, atn_cache=1024):
    env = Bvr(num_envs=1000)
    env.reset()
    tick = 0

    actions = [env.action_space.sample() for _ in range(atn_cache)]

    import time
    start = time.time()
    while time.time() - start < timeout:
        atn = actions[tick % atn_cache]
        env.step(atn)
        tick += 1

    print(f"SPS: {env.num_agents * tick / (time.time() - start)}")

if __name__ == "__main__":
    test_performance()
